Fix to support 0byte file up/download

begoldsm opened the pull request:
Previously this resulted in hangs in our uploader/downloader because the files would be added and then skipped over and never updated. This should fix that and ensure that zero byte files are properly honored.

begoldsm commented:
@brettcannon @lmazuel I have added you guys as reviewers. Can you take a look for sanity and make sure I am not missing anything? We need to add some regression tests for this scenario, which I am working on now.

lmazuel commented on azure/datalake/store/multithread.py:
```diff
@@ -468,14 +468,19 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
+                            nbytes += nwritten
+            else:
+                # need to explicitly flush the file in order to create it on the server when the file is empty.
+                fout.flush();
```
Are you sure it's needed? `Close` implies `flush` and `with` implies `close`. @brettcannon ?

lmazuel replied:
Hum, ok this is your `open` actually, not a Python one. I would prefer that your `open` respects the Python std and flush on close.

lmazuel replied:
Since your open is open in the public API, I'm pretty sure you will finally have the issue of someone expeting this behavior.

brettcannon replied:
Unneeded semi-colon

begoldsm replied:
fixed!

begoldsm replied:
We do have the close functionality built in for this which does do the flush. Here I want to be explicit though for readability. Additionally, if anything happened to the file that would cause the "close" operation to behave unexpectedly, I want to ensure that we tried to send the data to the server prior to close.

lmazuel commented on azure/datalake/store/multithread.py:
```diff
@@ -245,7 +245,7 @@ def get_chunk(adlfs, src, dst, offset, size, buffersize, blocksize,
             nbytes = 0
-            with closing(_fetch_range(adlfs.azure, src, start=offset,
+            with closing(_fetch_range_with_retry(adlfs.azure, src, start=offset,
                                       end=offset+size, stream=True)) as response:
```
Please re-indent after new method name

begoldsm replied:
reverted this change so indent should be ok.

lmazuel commented on azure/datalake/store/core.py:
```diff
@@ -733,8 +733,8 @@ def __exit__(self, *args):
-    if end <= start:
-        return b''
+    # if the caller gives a bad start/end combination, OPEN will throw and
+    # this call will bubble it up
```
LGTM

lmazuel commented on azure/datalake/store/multithread.py:
```diff
@@ -245,7 +245,7 @@ def get_chunk(adlfs, src, dst, offset, size, buffersize, blocksize,
         try:
             nbytes = 0
-            with closing(_fetch_range(adlfs.azure, src, start=offset,
+            with closing(_fetch_range_with_retry(adlfs.azure, src, start=offset,
```
Should not retries/delay/backoff be used in the the `_fetch_range_with_retry` call?

lmazuel replied:
Actually, you already in a retries loop, so now you retry inside of a retry (and this cannot be configured). It's what you want?

begoldsm replied:
Ah, good catch Laurent. I will revert this change, since we should not do nested retries here.

brettcannon commented on azure/datalake/store/multithread.py:
```diff
@@ -468,14 +468,19 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
-                    if nwritten:
-                        nbytes += nwritten
+            # For empty files there is no need to take the IO hit.
+            if size != 0:
```
Typically in Python you put the smaller branch at the top and the longer one at the bottom so it's easier to visually tell which block belongs to what statement.

begoldsm replied:
fair enough, fixed!

begoldsm replied:
Done!

brettcannon commented on azure/datalake/store/multithread.py:
```diff
@@ -468,14 +468,19 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
+                            return nbytes, None
+                        data = read_block(fin, o, miniblock, delimiter)
+                        nwritten = fout.write(data)
+                        if nwritten:
```
Is this necessary? Will `nwritten` ever be anything but an int? If it's always a number why not just unconditionally add the number to `nwritten`? (And then you can also just drop `nwritten` and and use `nbytes += fout.write(data)`.

begoldsm replied:
Looking at the code I would think that it will always return a number of bytes written, so long as BytesIO.write always returns an int or throws. I will make this change to clean this up either way since the logic doesn't really make sense to do the additional check.

lmazuel commented on azure/datalake/store/transfer.py:
```diff
@@ -269,6 +269,14 @@ def submit(self, src, dst, length):
         offsets = list(range(0, length, self._chunksize))
+
+        # in the case of empty files, ensure that the initial offset of 0 is properly added.
+        if len(offsets) == 0 and length == 0:
```
```python
if not offsets and length ==0
```

brettcannon replied:
This and the below guards could be simplified by having a top-level `if len(offsets) == 0` and then have an `if` inside that has guards of `not length` and `else`. E.g.
```python
if not len(offsets):
  if not length:
    offsets.append(0)
  else:
    raise ...
```

begoldsm replied:
Done!

lmazuel commented on azure/datalake/store/transfer.py:
```diff
@@ -269,6 +269,14 @@ def submit(self, src, dst, length):
+        if len(offsets) == 0 and length == 0:
+            offsets.append(0)
+
+        if len(offsets) == 0 and length != 0:
```
Please merge the two `if`
```python
if not offsets:
   if length == 0:
      offsets.append(0)
   else:
      raise DatalakeIncompleteTransferException("Suicide")

begoldsm replied:
done! Also I like your exception text 🗡 !

brettcannon commented on tests/test_multithread.py:
```diff
@@ -72,6 +72,21 @@ def create_remote_csv(fs, name, columns, colwidth, lines):
+        assert md5sum(fname) == checksum
+        assert os.stat(fname).st_size == size
+        assert linecount(fname) == lines
+        os.remove(fname)
```
Does this need to be guaranteed to run if there's an exception in ADLDownloader? If so you should use a `try`/`finally` block to make sure it happens if `fname` is actually created (or whatever pytest uses to register cleanup functions once a test finishes).

begoldsm replied:
I suppose I can be a good steward and do my best to cleanup after myself.

brettcannon replied:
:) I bring it up as I have found it easy to accidentally check in test files from test suites and not notice until later.

lmazuel commented on azure/datalake/store/utils.py:
```diff
@@ -86,7 +86,6 @@ def read_block(f, offset, length, delimiter=None):
         f.seek(start + length)
         seek_delimiter(f, delimiter, 2**16)
         end = f.tell()
-        eof = not f.read(1)
```
LGTM to me, seems not used. No side effect of read? (like moving any pointer). Didn't look at the entire code to check.

begoldsm replied:
No side effect that I am aware of. I think this was trying to confirm again that the full chunk was traversed, and it was using the broken logic of read that returned an empty byte array. Since that is all it was doing, there should be no side effect. 

I will be adding some more in depth tests for delimiter validation in the future.

lmazuel commented on azure/datalake/store/utils.py:
```diff
@@ -110,10 +109,14 @@ def seek_delimiter(file, delimiter, blocksize):
     """
-
+    # if we are at either the very beginning or very end of the file, 
+    # there is nothing to return so just return
```
Trust you :)

begoldsm replied:
Based on the logic as it is described here, I believe this to be true. The problem is that I am not entirely confident in any of the delimiter logic, however it is not currently hooked up, and I will spend more time validating this in the future.

The comment here is to explain the existing logic, as I did not change the existing logic. Comments are very helpful for maintainability :)

lmazuel commented on azure/datalake/store/multithread.py:
```diff
@@ -469,18 +469,17 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
-            if size != 0:
+            if size == 0:
+                # need to explicitly flush the file in order to create it on the server when the file is empty.
+                fout.flush()
```
I still think your close should flush and this is useless and counterproductive. Looking at your close implementation, there is a flush:
https://github.com/Azure/azure-data-lake-store-python/blob/master/azure/datalake/store/core.py#L706

If you miss the flush for an empty, I suggest to fix ADLFile instead.

begoldsm commented:
@lmazuel I removed the explicit flush since you are right :). I also removed all delimiter support and fixed the delimiter logic and tests and added a bunch of comments and TODOs around it, as well as this enchancement issue:
https://github.com/Azure/azure-data-lake-store-python/issues/132
